http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/sam/SamRetry.scala

```diff
@@ -0,0 +1,67 @@
+package org.broadinstitute.dsde.workbench.leonardo.dao.sam
+
+import cats.effect.Async
+import cats.mtl.Ask
+import org.broadinstitute.dsde.workbench.RetryConfig
+import org.broadinstitute.dsde.workbench.client.sam.ApiException
+import org.apache.http.HttpStatus
+import org.broadinstitute.dsde.workbench.util2.addJitter
+
+import java.net.SocketTimeoutException
+import scala.concurrent.duration._
+import org.broadinstitute.dsde.workbench.google2.tracedRetryF
+import org.broadinstitute.dsde.workbench.model.TraceId
+import org.typelevel.log4cats.StructuredLogger
+
+object SamRetry {
+
+  /**
+   * Initial delay of 1 second, exponential retry up to 5 times.
+   */
+  private val defaultSamRetryConfig =
+    RetryConfig(addJitter(1 seconds, 1 seconds), _ * 2, 5, isRetryable)
+
+  /**
+   * Requests made through the Sam client library sometimes fail with timeouts, generally due to
+   * transient network or connection issues. When this happens, the client library will throw an API
+   * exceptions with status code 0 wrapping a SocketTimeoutException. These errors should always be
+   * retried.
+   */
+  private def isTimeoutException(apiException: ApiException): Boolean =
+    (apiException.getCode == 0) && apiException.getCause.isInstanceOf[SocketTimeoutException]
+
+  private def isRetryable(throwable: Throwable): Boolean = throwable match {
+    case e: ApiException =>
+      isTimeoutException(e) ||
+      e.getCode == HttpStatus.SC_INTERNAL_SERVER_ERROR ||
+      e.getCode == HttpStatus.SC_BAD_GATEWAY ||
+      e.getCode == HttpStatus.SC_SERVICE_UNAVAILABLE ||
+      e.getCode == HttpStatus.SC_GATEWAY_TIMEOUT
+    case _ => false
+  }
+
+  /**
+   * Retries an effect with a given retry configuration and logging.
+   * Uses `tracedRetryF` from workbench-libs under the hood.
```

Comment:
One thing I want to evaluate is whether `tracedRetryF` suits our Sam needs here, especially the logging piece. It was originally used for GCP clients (but should be pretty general purpose).

Reply:
Yeah the retries work OK but I don't like the logging behavior that `tracedRetryF` gives us. It seems to always log the response which is not always desired (Sam responses can include things like tokens and keys).

Will explore more here..

Reply:
Fixed in https://github.com/DataBiosphere/leonardo/pull/4779/commits/1bc0d4520fb4c6436c106a7134bffd0f2516001f, not using the wb-libs logging anymore.